### PR TITLE
Upgrades to support presentations required by Eco

### PR DESCRIPTION
**Pull request by Britefury**
Implemented border element for visual guides and flair.
Added hacked style sheet implementation to JSON viewer tool.

### CODE REVIEW

File: src/lib/lspace/elements/border_element.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    /// Update element Y requisition
+    fn element_update_y_req(&self, y_req: &LReq) -> bool {
+        return self.m.borrow_mut().req.update_y_req(y_req);
+    }
+
+    /// Paint the element content that is contributed by the element itself
+    fn draw_self(&self, cairo_ctx: &Context, visible_region: &BBox2) {
+        let mm = self.m.borrow();
+        let alloc = &mm.alloc;
+        self.border.draw_background(cairo_ctx,
```
Comment (ltratt) [inline src/lib/lspace/elements/border_element.rs:94]: The indent of this call's parameters doesn't match the call immediately below it.

File: src/lib/lspace/graphics/border.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                Border::border_path(cairo_ctx, x, y, w, h, r);
+            }
+        };
+    }
+
+    pub fn draw(&self, cairo_ctx: &Context, x: f64, y: f64, w: f64, h: f64) {
+        match self {
+            &Border::SolidBorder{thickness: t, rounding: r, border_colour: col, ..} => {
+                let prev_width = cairo_ctx.get_line_width();
+//                let prev_source = cairo_ctx.get_source();
```
Comment (ltratt) [inline src/lib/lspace/graphics/border.rs]: Remove commented line.

File: src/lib/lspace/graphics/border.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+//                cairo_ctx.set_source(&*prev_source);
+            },
+            _ => {}
+        };
+    }
+
+    pub fn draw_background(&self, cairo_ctx: &Context, x: f64, y: f64, w: f64, h: f64) {
+        match self {
+            &Border::SolidBorder{thickness: t, rounding: r, background_colour: Some(col), ..} => {
+//                let prev_source = cairo_ctx.get_source();
```
Comment (ltratt) [inline src/lib/lspace/graphics/border.rs]: Remove commented line.

File: src/lib/lspace/graphics/border.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+                cairo_ctx.new_path();
+                Border::border_path(cairo_ctx, x + t * 0.5, y + t * 0.5, w - t, h - t, r);
+                cairo_ctx.set_source_rgba(col.r as f64, col.g as f64, col.b as f64, col.a as f64);
+                cairo_ctx.fill();
+
+//                cairo_ctx.set_source(&*prev_source);
```
Comment (ltratt) [inline src/lib/lspace/graphics/border.rs]: Remove commented line.

File: src/lib/lspace/graphics/border.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                cairo_ctx.set_source_rgba(0.0, 0.0, 0.0, 1.0);
+            },
+            &Border::FilledBorder{rounding: r, background_colour: Some(col), ..} => {
+//                let prev_source = cairo_ctx.get_source();
```
Comment (ltratt) [inline src/lib/lspace/graphics/border.rs]: Remove commented line.

File: src/lib/lspace/graphics/border.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+                cairo_ctx.new_path();
+                Border::border_path(cairo_ctx, x, y, w, h, r);
+                cairo_ctx.set_source_rgba(col.r as f64, col.g as f64, col.b as f64, col.a as f64);
+                cairo_ctx.fill();
+
+//                cairo_ctx.set_source(&*prev_source);
```
Comment (ltratt) [inline src/lib/lspace/graphics/border.rs]: Remove commented line.

**Comment by Britefury** [comment]
Should all be fixed.

**Comment by ltratt** [comment]
Looks good. Do you want to do any squashing?

**Comment by Britefury** [comment]
Yes, sure.

**Comment by Britefury** [comment]
Ready to merge.

**Pull request merged**